# Variables

### Clicker #0

Please answer this iClicker question. 

- a) okay, sure
- b) yes, of course!
- c) I don't want to
- d) why are we doing this
- e) I hate iClickers

## Programming With Python

Programming: a way to ask computer to store values (variables), and do things with them (operations).

### Variables and Comments

In [ ]:
# This is a comment. You can write a comment by using a `#`
# A comment is a note for someone reading your code
# It doesn't get executed by the kernel

one_dozen = 12  # on the left of '=' is the variable name, on the right of '=' is the value stored in the variable.  
# To the right of # is a comment

bakers_dozen = 13 # Comments can be 'inline', like this one, or by themselves like the one at the top of this cell

darth_vader = "Luke's Father" # variables can be called anything you want, comments help people understand your code

## Declaring Variables

<div class="alert alert-success">
In programming, variables are things that store values. Variables are declared with `name = value`.
</div>

## The rules of declaring variables

- Names are always on the left of the `=`, values are always on the right
- Names are case sensitive
- Variables must start with letters
    - After that, they can include numbers, and underscores
    - They cannot include special characters (like &, *, #, etc)
- Python doesn't care what you name you variables
    - Humans do care. Pick names that describe the data or value that they store.
    - Make it so you don't have to explain your code to anyone.
        - They should just get it from reading the code.

## Spelling and grammar count.  

- Most errors when we're starting out are because of typos and indents. 
- Programming languages are unforgiving.
    - They don't care how close you are
    - They don't give you the benefit of the doubt
    - Right or wrong, they follow the instructions you type exactly
    - It is your responsibility as a coder to make sure your code is free from typos.


Every experiment we do in Psychology will have lots of variables.  Let's start with our mock participant in our mock experiment. After providing informed consent to an experiment participants are assigned an ID number to accompany their data instead of their names to provide anonymity.  In the example below `participant` is a variable that will contain their ID number. Remember that in jupyter notebooks, when you type the name of a variable, it displays the value. 

In [1]:
participant = 1331
participant

1331

Notice that I used lower-cased letters for `participant`.

- In Python, you should always use lower-cased letters for the names of youre variables. 
- This helps other people who read your code quickly identify the type of object they are looking at.
- If you need to make a variable that is more than one word, use an underscore `_` instead of a space.
- If you want to learn the rules of formatting your code, google "Python PEP-8"


####  You can assign a variable to another variable

In [2]:
participant_number = participant
participant_number

1331

You can change the value stored in variables

In [3]:
participant += 1  # this means participant = participant + 1
participant

1332

This will also change the variable we assigned it to.  That's because when we wrote `participant_number = participant`, we didn't copy the value stored in `participant`, i.e. `1331`, instead, we pointed at the variable, so when `participant` changes, so does `participant_number`.

In [4]:
participant_number

1331

But when we change participant_number, it doesn't affect participant.  It's not a 2-way street.

In [ ]:
participant_number += 1
print('participant_number: ', participant_number)
print('participant: ', participant)

### Clicker #1

What do you think the value of `participant_number` will be now if we add 1 to `participant` by executing the following code: 
`participant += 1`?

- a) 1331
- b) 1332
- c) 1333
- d) I'm totally lost, what's going on?
- e) I hate iClickers, please stop doing this.

In [5]:
participant += 1
print('participant_number: ', participant_number)
print('participant: ', participant)

participant_number:  1331
participant:  1333


So now `participant_number` and `participant` store different values. That can make things really weird unless you are doing this on purpose. Practical uses of this are keeping track of multiple items... but be careful and make sure you code this with intent. It's not really for beginners.

Your variables don't have to be about participants.  They can and should be about whatever they represent.  For math and graphing, we often use the single letters `x` and `y`.

In [6]:
x = 1
y = 2
x + y

3

You can use numbers and letters, but variables have to start with a letter.

In [7]:
r2d2 = 'white and blue robot'
c3p0 = 'gold robot'
print(r2d2,'\n', c3p0)

white and blue robot 
 gold robot


# Print

- The print function outputs things to the screen, not the printer
- When we run code in Jupyter notebook, it only shows the final output, like in the previous slide
- When we want to see more, or want to write code for use outside of Jupyter Notebooks, use "print"

## You can print the values stored in variables

In [8]:
apple = "Honey Crisp"
pi = 3.14 

In [9]:
print(apple)

Honey Crisp


In [10]:
print(pi)

3.14


In [11]:
print (apple, pi)

Honey Crisp 3.14


## You can print text

In [12]:
print("To be, or not to be, that is the question:\n" 
      "Whether 'tis nobler in the mind to suffer\n"
      "The slings and arrows of outrageous fortune,\n"
      "Or to take Arms against a Sea of troubles,\n"
      "And by opposing end them")


To be, or not to be, that is the question:
Whether 'tis nobler in the mind to suffer
The slings and arrows of outrageous fortune,
Or to take Arms against a Sea of troubles,
And by opposing end them


<div class="alert alert-info">
You can see I typed "\n" to tell the computer to make a new line.\ is called an escape string.  Google 'python escape strings' to learn more escape strings.
    </div>

## You can print text and variables together

In [13]:
r_squared = 0.76

# This defines another variable
effect_size = 'large'

print("The r^2 value is: ", r_squared, "\n", "The effect size is ", effect_size)

The r^2 value is:  0.76 
 The effect size is  large


## You can print a lot of things a lot of ways
We'll learn more about string formatting and string manipulations later.

# <font color="red">Warning - "print" is different in Python 3.X and Python 2.7</font>



## A lot of web help for Python is in Python 2.7
- In Python 2.7, which we don't use in class, you print like this:
    - `print "lorum ipsum"`
- In Python 3.X, the version we use in class, you print like this:
    - `print("lorum ipsum")`
- To fix the print statements, just make sure you use parentheses like this `print("this way")`.

# Back to Variables....

## Code Variables != Math Variables

In mathematics: `=` refers to equality (as a statement of truth).

In coding: `=` refers to assignment. You can read the '=' as 'gets'.  

x = 2 + 5 can be read as "x gets the value that results when you add 2 plus 5".

In coding, you can use `==` to evaluate equality.

Math: What is x?

`102 = 10x + 2`

Code: What is x?

`x = x + 1`

### Clicker #2

After executing the following code, what will be the value of `my_var`?

In [14]:
my_var = 2

my_var = my_var + 1

print(my_var)

3


- a) 2
- b) 3
- c) "my_var + 1"
- d) This code will fail
- e) I'm totally lost.  What is going on here?


### Reminder

- In programming `=` means assignment
- Anything to the right of the `=` is evaluated before assignment

## Namespace

<div class="alert alert-success">
The namespace is the 'place' where all your currently defined code is declared - all the things you have stored in active memory. 
</div>

In [15]:
# You can list everything declated in the namespace with '%whos'
%whos

Variable             Type     Data/Info
---------------------------------------
apple                str      Honey Crisp
c3p0                 str      gold robot
effect_size          str      large
my_var               int      3
participant          int      1333
participant_number   int      1331
pi                   float    3.14
r2d2                 str      white and blue robot
r_squared            float    0.76
x                    int      1
y                    int      2


## <center> Don't reuse & recycle variable names</center>
<img src="img/recycle.jpg">


----------------
- Technically you can re-use variables.  
- It's bad to re-use variable names because it's easy to get mixed up. 
- If you make a variable by accident, or have a type, or just don't want it anymore, just leave it alone.  
    - It's not going to hurt anyone if you just don't type it.
- I reuse variable names in class to help teach.
    - Do as I say, not as I do.

### iClicker Question #2

After executing the following code, what will be the value of `var_2`?

In [ ]:
var_2 = var_1 = 1

print(var_2)

- a) 'var_1'
- b) 1
- c) 2
- d) This code will fail

### Reminders

- There can be more than one assignment in a single line
- Anything to the right of the `=` is evaluated before assignment
    - This process proceeds from right to left

# Variable Types

<div class="alert alert-success">
Every variable has a 'type', which refers to the kind of variable that it is, and how the computer stores that data.
</div>

You don't have to tell Python what type of variable something is.  It figures it out itself by the nature of the contents. In other programming languages, you do have to specify this when declaring variables, so we are happy we are using Python so we can bypass this step.

In [16]:
# Declare a variable
variable_name = 1

# You can always ask Python 'what type is this variable' using:
type(variable_name)

int

### Int

<div class="alert alert-success">
<code>Integers</code> store whole numbers.
</div>

In [17]:
my_integer = 1
another_integer = 321
a_negative_integer = -345
zero = 0

### Float

<div class="alert alert-success">
<code>Floats</code> store signed, decimal-point numbers.  </div>
The decimal point 'floats' around, depending on how many decimal places you need.
They don't have a signature, they are signed (i.e.) positive (+) or negative (-).


In [18]:
my_float = 1.0
another_float = -231.45

### String

<div class="alert alert-success">
<code>Strings</code> store characters, as text.  
</div>


In [19]:
my_string = 'words, words, words'
another_string = 'more words'

# Note that strings can be defined with either '' or ""
and_another = "and some more"

<div class="alert alert-success">
<code>Strings<code> can also store numbers, as text.  
</div>

In [20]:
one = "1"
two = "2"
one + two

'12'

## Boolean

<div class="alert alert-success">
    <code>Booleans</code> store <code>True</code> or <code>False</code>. 
</div>
When a variable is:
- <code>False</code>
    - empty string (e.g. <code>''</code>)
    - or <code>0</code>
        - Python recognizes it as False.
- <code>True</code>
    - empty string (e.g. <code>''</code>)
    - or not <code>0</code>
        - Python recognizes it as True.



## iClicker Question #3


When a variable is `'False'`

e.g. `my_var = 'False'`

Does Python recognize this variable as   
1) `True`   
2) `False`   
3) I'm lost, what's going on?  

In [21]:
# iClicker #3 solution
my_var = 'False'

In [ ]:
my_var is True

In [ ]:
my_var is False

In [ ]:
my_var is 'False'

## None

<div class="alert alert-success">
<code>None</code> is a special type that stores <code>None</code>, used to denote a null or empty value.
</div>

In [1]:
the_concept_of_nothing = None

In [23]:
the_concept_of_nothing is True

False

In [25]:
the_concept_of_nothing is False

False

In [26]:
the_concept_of_nothing is None

True

That's deep.

### Reminders

- Multiple variables can relate to the same value(s)
    - e.g. `x = 3` & `y = 3`can both peacefully co-exist

# Operators

## Assignment Operators

Assignment operators are used in Python to assign values to variables.

a = 5 is a simple assignment operator that assigns the value 5 on the right to the variable a on the left.

There are various compound operators in Python like a += 5 that adds to the variable and later assigns the same. It is equivalent to a = a + 5.

<div class="alert alert-success">
    Python uses <code>=</code> for assignment. This might sound straightforward, but other programming langauges like <code>R</code> use <code>&lt;-</code> for assignment. Rejoice that we are learning how to code using Python as opposed to R.
</div>

- In Python you can read the `=` sign as "gets".
- To compare things, you can read `==` as "is equivalent to".
- `<-` is not a thing in Python.

<table border="1">
    <H2><center>Assignment operators in Python</center></H2>
	<tbody>
		<tr>
			<th>Operator</th>
			<th>Example</th>
			<th>Equivatent to</th>
		</tr>
		<tr>
			<td>=</td>
			<td>x = 5</td>
			<td>x = 5</td>
		</tr>
		<tr>
			<td>+=</td>
			<td>x += 5</td>
			<td>x = x + 5</td>
		</tr>
		<tr>
			<td>-=</td>
			<td>x -= 5</td>
			<td>x = x - 5</td>
		</tr>
		<tr>
			<td>*=</td>
			<td>x *= 5</td>
			<td>x = x * 5</td>
		</tr>
		<tr>
			<td>/=</td>
			<td>x /= 5</td>
			<td>x = x / 5</td>
		</tr>
		<tr>
			<td>%=</td>
			<td>x %= 5</td>
			<td>x = x % 5</td>
		</tr>
		<tr>
			<td>//=</td>
			<td>x //= 5</td>
			<td>x = x // 5</td>
		</tr>
		<tr>
			<td>**=</td>
			<td>x **= 5</td>
			<td>x = x ** 5</td>
		</tr>
		<tr>
			<td>&amp;=</td>
			<td>x &amp;= 5</td>
			<td>x = x &amp; 5</td>
		</tr>
		<tr>
			<td>|=</td>
			<td>x |= 5</td>
			<td>x = x | 5</td>
		</tr>
		<tr>
			<td>^=</td>
			<td>x ^= 5</td>
			<td>x = x ^ 5</td>
		</tr>
		<tr>
			<td>&gt;&gt;=</td>
			<td>x &gt;&gt;= 5</td>
			<td>x = x &gt;&gt; 5</td>
		</tr>
		<tr>
			<td>&lt;&lt;=</td>
			<td>x &lt;&lt;= 5</td>
			<td>x = x &lt;&lt; 5</td>
		</tr>
	</tbody>
</table>

https://www.programiz.com/python-programming/operators

In [27]:
my_var = 1

## Math Operators

<div class="alert alert-success">
Python uses mathematical operators `+`, `-`, `*`, `/` for 'sum', 'substract', 'multiply', and 'divide'. These operators return numbers.
</div>

<table border="1">
    <H2><center>Arithmetic operators in Python</center></H2>
	<tbody>
		<tr>
			<th>Operator</th>
			<th>Meaning</th>
			<th>Example</th>
		</tr>
		<tr>
			<td>+</td>
			<td>Add two operands or unary plus</td>
			<td>x + y<br />
				+2</td>
		</tr>
		<tr>
			<td>-</td>
			<td>Subtract right operand from the left or unary minus</td>
			<td>x - y<br />
				-2</td>
		</tr>
		<tr>
			<td>*</td>
			<td>Multiply two operands</td>
			<td>x * y</td>
		</tr>
		<tr>
			<td>/</td>
			<td>Divide left operand by the right one (always results into float)</td>
			<td>x / y</td>
		</tr>
		<tr>
			<td>%</td>
			<td>Modulus - remainder of the division of left operand by the right</td>
			<td>x % y (remainder of x/y)</td>
		</tr>
		<tr>
			<td>//</td>
			<td>Floor division - division that results into whole number adjusted to the left in the number line</td>
			<td>x // y</td>
		</tr>
		<tr>
			<td>**</td>
			<td>Exponent - left operand raised to the power of right</td>
			<td>x**y (x to the power y)</td>
		</tr>
	</tbody>
</table>

https://www.programiz.com/python-programming/operators

Addition

In [28]:
2 + 3

5

Subtraction

In [29]:
2 - 3


-1

Multiplication

In [30]:
2 * 3

6

Division

In [31]:
2 / 3

0.6666666666666666

<div class="alert alert-danger">
Remember that Python uses <code>/</code> for divisions, and <code>%</code> for remainder (called modulus). These also return numbers.
</div>

Division

In [ ]:
3 / 2

Modulus

In [ ]:
3 % 2

<div class="alert alert-danger">
Remember that Python uses <code>**</code> exponentiation. <code>^</code> is for binary comparisons.  We won't be doing that in this course, so for now, you can safely never use <code>^</code>. Also, Python doesn't have a built-in square root function.  Remember your math and $\sqrt{X}$ is the same as $X^{0.5}$. Yes, there are square root functions available but for now we are learning python's core programming fundamentals.  We'll learn how to <code>import</code> code later.</div>

In [32]:
5 ** 0

1

In [33]:
81 ** 0.5

9.0

### Don't forget you can do math on variables, and you can also store the results as variables.

In [2]:
slice = 1
pizza = slice / 8
print(pizza)

0.125


### If you do math on non-numeric data types, funky things can happen....

In [3]:
# Recall the Beatles White Album...
revolution = "#9"
revolution * 1000

'#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#9#

## Boolean Logic

<div class="alert alert-success">
Python has `and`, `or` and `not` for boolean logic. These operators return booleans.
</div>

<table border="1">
	<H2><center>Logical operators in Python</center></H2>
	<tbody>
		<tr>
			<th>Operator</th>
			<th>Meaning</th>
			<th>Example</th>
		</tr>
		<tr>
			<td>and</td>
			<td>True if both the operands are true</td>
			<td>x and y</td>
		</tr>
		<tr>
			<td>or</td>
			<td>True if either of the operands is true</td>
			<td>x or y</td>
		</tr>
		<tr>
			<td>not</td>
			<td>True if operand is false (complements the operand)</td>
			<td>not x</td>
		</tr>
	</tbody>
</table>

In [4]:
x = True
y = False

In [5]:
x and y

False

In [6]:
x or y

True

In [7]:
not x

False

In [8]:
not y

True

### Clicker #1

What will the following boolean expression evaluate as:

In [ ]:
x and not y or x

- a) True
- b) False
- c) None
- d) This code will fail

## Comparison Operators

Python has comparison operators `==`, `!=`, `<`, `>`, `<=`, and `>=` for value comparisons. These operators return booleans.

<table border="1">
	<H2><center>Comparision operators in Python</center></H2>
	<tbody>
		<tr>
			<th>Operator</th>
			<th>Meaning</th>
			<th>Example</th>
		</tr>
		<tr>
			<td>&gt;</td>
			<td>Greater that - True if left operand is greater than the right</td>
			<td>x &gt; y</td>
		</tr>
		<tr>
			<td>&lt;</td>
			<td>Less that - True if left operand is less than the right</td>
			<td>x &lt; y</td>
		</tr>
		<tr>
			<td>==</td>
			<td>Equal to - True if both operands are equal</td>
			<td>x == y</td>
		</tr>
		<tr>
			<td>!=</td>
			<td>Not equal to - True if operands are not equal</td>
			<td>x != y</td>
		</tr>
		<tr>
			<td>&gt;=</td>
			<td>Greater than or equal to - True if left operand is greater than or equal to the right</td>
			<td>x &gt;= y</td>
		</tr>
		<tr>
			<td>&lt;=</td>
			<td>Less than or equal to - True if left operand is less than or equal to the right</td>
			<td>x &lt;= y</td>
		</tr>
	</tbody>
</table>

https://www.programiz.com/python-programming/operators

In [ ]:
x == x

In [ ]:
x == y

In [ ]:
'aa' == 'aa'

In [ ]:
12 >= 13

## Comparing Identity Operators

<div class="alert alert-success">
Python uses `is` and `is not` to compare identity. These operators return booleans.
</div>

<table border="1">
    <H2><center>Identity operators in Python</center></H2>
	<tbody>
		<tr>
			<th>Operator</th>
			<th>Meaning</th>
			<th>Example</th>
		</tr>
		<tr>
			<td>is</td>
			<td>True if the operands are identical (refer to the same object)</td>
			<td>x is True</td>
		</tr>
		<tr>
			<td>is not</td>
			<td>True if the operands are not identical (do not refer to the same object)</td>
			<td>x is not True</td>
		</tr>
	</tbody>
</table>

https://www.programiz.com/python-programming/operators

In [9]:
a = 927
b = a
c = 927

In [11]:
print(a is b)
print(c is a)
print(a is not c)
print(a == c)

True
False
True
True


In [12]:
a == b == c

True

<h3><a id="membership" name="membership"></a>Membership operators</h3>

<p><code>in</code> and <code>not in</code> are the membership operators in Python. They are used to test whether a value or variable is found in a sequence (<code>string</code>, <code>list</code>, <code>tuple</code>, <code>set</code>, and <code>dictionary</code>). In a dictionary we can only test for presence of key, not the value.</p>



<table border="1">
	<tbody>
		<tr>
			<th>Operator</th>
			<th>Meaning</th>
			<th>Example</th>
		</tr>
		<tr>
			<td>in</td>
			<td>True if value/variable is found in the sequence</td>
			<td>5 in x</td>
		</tr>
		<tr>
			<td>not in</td>
			<td>True if value/variable is not found in the sequence</td>
			<td>5 not in x</td>
		</tr>
	</tbody>
</table>

https://www.programiz.com/python-programming/operators

In [13]:
name = "Feinberg"

'be' in name

True

In [14]:
'elbow' not in name

True

## String Operations

<div class="alert alert-success">
Operators sometimes do different things on different data. For example, `+` on strings does concatenation.
</div>

In [15]:
'a' + 'b' + 'c'

'abc'

In [16]:
one = "1"
two = "2"
one + two

'12'

In [17]:
first_name = "David"
space = " "
last_name = "Feinberg"
full_name = first_name + space + last_name
full_name

'David Feinberg'

### You can't add numbers and strings

In [18]:
full_name + 7

TypeError: can only concatenate str (not "int") to str

### ... unless the numbers are stored as strings

In [19]:
full_name + '7'

'David Feinberg7'

### But, remember our Revolution #9 example... you can multiply strings

In [20]:
full_name * 21

'David FeinbergDavid FeinbergDavid FeinbergDavid FeinbergDavid FeinbergDavid FeinbergDavid FeinbergDavid FeinbergDavid FeinbergDavid FeinbergDavid FeinbergDavid FeinbergDavid FeinbergDavid FeinbergDavid FeinbergDavid FeinbergDavid FeinbergDavid FeinbergDavid FeinbergDavid FeinbergDavid Feinberg'

### Good thing you can't divide strings also.   What would that even look like?

In [21]:
full_name / 21

TypeError: unsupported operand type(s) for /: 'str' and 'int'

## Chaining Operators

<div class="alert alert-success">
Operators and variables can also be chained together into arbitrarily complex expressions.
</div>

In [22]:
 ('PNB' + " " + '2A03' == 'PNB 2A03')

True

### You can do that with variables too...

In [23]:
department = "PNB"
space = " "
course_code = "2A03"

department + space + course_code == 'PNB 2A03'

True

### Note that you can use parentheses to chunk sections

In [ ]:
(13 % 7 >= 7) and (department + space + course_code == 'PNB 2A03')

### Clicker #2

What will the following expression evaluate as:

In [ ]:
2 ** 2 >= 4 and 13 % 3 > 1

- a) True
- b) False
- c) None
- d) This code will fail

In [ ]:
The end